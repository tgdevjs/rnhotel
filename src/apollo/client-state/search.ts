import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

// typeDefs
const typeDefs = gql`
  extend type Query {
    currentSearch: Search
  }

  extend type Mutation {
    setCurrentSearch(input: searchInput!): Search
  }

  extend input searchInput {
    endDay: String
    startDay: String
  }

  extend type Search {
    endDay: String
    startDay: String
  }
`;

// defaults
const defaults = {
  currentSearch: {
    __typename: 'Search',
    endDay: null,
    startDay: null,
  },
};

// queries
export const currentSearchQuery = gql`
  query getCurrentSearch {
    currentSearch @client {
      endDay
      startDay
    }
  }
`;

// mutations
const setCurrentSearchMutation = gql`
  mutation setCurrentSearch($input: searchInput) {
    setCurrentSearch(input: $input) @client
  }
`;

// resolvers
const setCurrentSearch = (obj, { input }, { cache }) => {
  const { currentSearch } = cache.readQuery({ query: currentSearchQuery });
  const data = {
    currentSearch: {
      ...currentSearch,
      ...input,
    },
  };

  cache.writeQuery({
    query: currentSearchQuery,
    data,
  });
};

const resolvers = {
  Mutation: {
    setCurrentSearch,
  },
};

export const clientState = {
  typeDefs,
  defaults,
  resolvers,
};

// enhancers
const currentSearchQueryHandler = ({ data }) => ({ ...data.currentSearch });

export const withSearchQuery = graphql(currentSearchQuery, { props: currentSearchQueryHandler });
export const withSearchMutation = graphql(setCurrentSearchMutation, { name: 'setCurrentSearch' });
