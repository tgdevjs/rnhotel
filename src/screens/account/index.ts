export * from './Account';
export * from './AccountGuest';
export * from './AccountMenu';
export * from './JoinModal';
export * from './SignInModal';
