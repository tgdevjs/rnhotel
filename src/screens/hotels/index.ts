export * from './Hotels';
export * from './ConfirmReservation';
