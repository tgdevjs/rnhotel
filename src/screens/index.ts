export * from './account';
export * from './hotels';
export * from './search';
export * from './stays';
