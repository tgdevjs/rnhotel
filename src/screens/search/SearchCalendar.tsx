import React, { Component } from 'react';
import { Button } from 'react-native';
import PropTypes from 'prop-types';

import { Calendar } from '../../components';

export class SearchCalendar extends Component {
  static navigationOptions = ({ navigation }) => {
    const { startDay, endDay, onSelectDates } = navigation.state.params;

    return {
      headerRight: (
        <Button
          onPress={() => {
            onSelectDates({ startDay, endDay });
            navigation.goBack();
          }}
          disabled={!endDay}
          title="Done"
        />
      ),
    };
  };

  onSetDates = dates => {
    const {
      navigation: { setParams },
    } = this.props;

    setParams(dates);
  };

  render() {
    const {
      navigation: {
        state: {
          params: { endDay, startDay },
        },
      },
    } = this.props;

    return <Calendar endDay={endDay} onSetDates={this.onSetDates} startDay={startDay} />;
  }
}

SearchCalendar.propTypes = {
  navigation: PropTypes.shape({}).isRequired,
};
