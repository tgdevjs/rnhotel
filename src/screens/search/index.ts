export * from './Search';
export * from './SearchCalendar';
