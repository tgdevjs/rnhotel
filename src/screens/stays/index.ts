export * from './Stays';
export * from './StayItem';
export * from './StaysList';
export * from './StaysListGuest';
