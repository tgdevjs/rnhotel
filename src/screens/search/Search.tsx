import React from "react";
import { Button, StyleSheet, Text, TouchableOpacity, View } from "react-native";
import { NavigationScreenProp } from "react-navigation";
import { compose } from "react-apollo";

import { DateRange, SignInButton } from "../../components";
import {
  withSearchQuery,
  withSearchMutation
} from "../../apollo/client-state/search";
import { withUserQuery } from "../../apollo/client-state/user";
import { UserType } from "../../types";

type Props = {
  endDay: string;
  navigation: NavigationScreenProp<any, any>;
  setCurrentSearch: () => void;
  startDay: string;
  user: UserType;
};
const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    flex: 1
  },
  signInSection: {
    flexDirection: "row"
  },
  signInContainer: {
    flex: 1,
    alignItems: "flex-start",
    margin: 20
  },
  userText: {
    fontSize: 16
  },
  main: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center"
  },
  welcome: {
    fontSize: 20,
    textAlign: "center",
    margin: 10
  },
  instructions: {
    textAlign: "center",
    color: "#333333",
    marginBottom: 5
  },
  buttonContainer: {
    marginTop: 30
  }
});

export class Search extends React.Component<Props> {
  static navigationOptions = () => ({ headerTitle: "Search" });

  onSelectDates = date => {
    const { setCurrentSearch } = this.props;

    setCurrentSearch({ variables: { input: { ...date } } });
  };

  render() {
    const {
      endDay,
      navigation: { navigate },
      startDay,
      user: { name }
    } = this.props;
    const { onSelectDates } = this;

    return (
      <View style={styles.container}>
        <View style={styles.signInSection}>
          <View style={styles.signInContainer}>
            {name ? (
              <Text style={styles.userText}>Hi, {name} </Text>
            ) : (
              <SignInButton onPress={() => navigate("SignIn")} />
            )}
          </View>
        </View>
        <View style={styles.main}>
          <TouchableOpacity
            style={{ width: 400 }}
            onPress={() => {
              navigate("Calendar", {
                endDay,
                onSelectDates,
                startDay
              });
            }}
          >
            <DateRange startDay={startDay} endDay={endDay} />
          </TouchableOpacity>
          <View style={styles.buttonContainer}>
            <Button title="FindHotels" onPress={() => navigate("Hotels")} />
          </View>
        </View>
      </View>
    );
  }
}

export const SearchWithQuery = compose(
  withSearchQuery,
  withSearchMutation,
  withUserQuery
)(Search);
