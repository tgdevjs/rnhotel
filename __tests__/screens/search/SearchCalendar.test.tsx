import React from "react";
import { SearchCalendar } from "../../../src/screens/search";

function searchCalendar(props = {}) {
  const defaultProps = {
    navigation: {
      setParams: jest.fn(() => null),
      state: { params: { endDay: "2019-06-03", startDay: "2019-06-01" } }
    },
    ...props
  };
  return shallow(<SearchCalendar {...defaultProps} />);
}

describe("SearchCalendar", () => {
  it("renders default", () => {
    expect(searchCalendar()).toMatchSnapshot();
  });
});
