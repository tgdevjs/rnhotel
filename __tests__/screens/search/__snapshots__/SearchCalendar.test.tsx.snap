// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchCalendar renders default 1`] = `
<Calendar
  endDay="2019-06-03"
  onSetDates={[Function]}
  startDay="2019-06-01"
/>
`;
